src/client/renderingengine.cpp

```diff
@@ -138,7 +134,27 @@ RenderingEngine::RenderingEngine(IEventReceiver *receiver)
 	params.OGLES2ShaderPath = (porting::path_share + DIR_DELIM + rel_path + DIR_DELIM).c_str();
 #endif
 
-	m_device = createDeviceEx(params);
+	if (driverType != video::EDT_NULL) {
+		m_device = createDeviceEx(params);
+		if (!m_device)
+			errorstream << "Failed to initialize the device using the " << RenderingEngine::getVideoDriverInfo(driverType).friendly_name << " driver." << std::endl;
+	}
+	if (!m_device) {
+		// try to find any working video driver
+		for (auto driverTypeToTry: drivers) {
+			if (driverTypeToTry == video::EDT_NULL || driverTypeToTry == driverType)
+				continue;
+			params.DriverType = driverTypeToTry;
+			m_device = createDeviceEx(params);
+			if (m_device) {
+				driverType = driverTypeToTry;
+				errorstream << "Falling back to " << RenderingEngine::getVideoDriverInfo(driverType).friendly_name << "." << std::endl;
+				break;
+			}
+		}
+	}
+	if (!m_device)
+		throw std::runtime_error("Could not initialize the device with any supported video driver");
```

Comment:
This part would be good to get in; without it the process crashes on the next line when `createDeviceEx()` fails.

I might throw another type of exception because `runtime_error` results in dumping a core, which is useless for this error and a bit severe. Maybe the existing `ClientStateError` or roll your own video-specific exception type?

Reply:
`ClientStateError` is *definitely* wrong choice, it is used in the server to denote that there is something wrong with the client. And, what’s the point in using another exception type if no exceptions are caught up the call stack anyway?

Reply:
> And, what’s the point in using another exception type if no exceptions are caught up the call stack anyway?

As I said, left uncaught this exception can result in a core file, which while useful for debugging in general, doesn't have much use here. It's just nicer to not litter the user's machine with such files for lack of a more appropriate exception type, but I guess there are lots of ways to disable core files at the OS level so I can deal with it. It's just a good code hygiene suggestion.